README.md

```diff
@@ -10,12 +10,328 @@ A core widget library for Dojo 2.
 
-To see how the widgets are utilised, there is an [example application](https://github.com/kitsonk/app-example) available.  Currently, outside of the source distribution itself, there is not a lot of documentation, as the code base is under heavy development.
+The examples below are provided in TypeScript syntax. The package does work under JavaScript, but for clarity, the examples will only include one syntax.
+
+### Base Widget
+
+The class `createWidgetBase` provides all base Dojo 2 widget functionality including caching and widget lifecycle management. It can be used directly or extended to create custom widgets.
+
+To customise the widget an optional `options` argument can be provided with the following interface.
+
+**Type**: `WidgetOptions<WidgetState>` - All properties are optional.
+
+|Property|Type|Description|
+|---|---|---|
+|id|string|identifier for the widget|
+|state|WidgetState|Initial state of the widget|
+|stateFrom|StoreObservablePatchable|Observable that provides state for the widget|
+|listeners|EventedListenersMap|Map of listeners for to attach to the widget|
+|tagName|string|Override the widgets     tagname|
+|getChildrenNodes|Function|Function that returns an array of children DNodes|
+|nodeAttributes|Function[]|An array of functions that return VNodeProperties to be applied to the VNode|
+
+By default the base widget class applies `id`, `classes` and `styles` from the widgets specified `state` (either by direct state injection or via an observable store).
+
+#### Simple Widgets
+To create a basic widget `createWidgetBase` can be used directly by importing the class.
+
+```ts
+import createWidgetBase from 'dojo-widgets/bases/createWidgetBase';
+
+const myBasicWidget = createWidgetBase();
+```
+
+Creates the following DOM element:
+
+```html
+<div></div>
+```
+The following example demonstrates how `id`, `classes` and `styles` are applied to the generated DOM.
+
+```ts
+import createWidgetBase from 'dojo-widgets/bases/createWidgetBase';
+
+const myBasicWidget = createWidgetBase({
+    state: {
+        id: 'my-widget',
+        classes: [ 'class-a', 'class-b' ],
+        styles: [ 'width:20px' ]
+    }
+});
+```
+
+Creates the following DOM element:
+
+```html
+<div data-widget-id="my-widget" class="class-a class-b" styles="width:20px"></div>
+```
+Alternatively state can be derived directly from an observable store passed via `stateFrom`. The following code will create the same DOM element as the above example.
+
+```ts
+const widgetStore = createObservableStore({
+    data: [
+        {
+            id: 'my-widget',
+            classes: [ 'class-a', 'class-b' ],
+            styles: [ 'width:20px' ]
+        }
+    ]
+});
+
+const myBasicWidget = createWidgetBase({
+    id: 'my-widget',
+    stateFrom: widgetStore
+});
+```
+
+#### `d`
+
+`d` is a function that is used within Dojo 2 to express widget hierarchical structure using both Dojo 2 widget factories or Hyperscript, it is imported via
```

Comment:
My concern with calling everything Dojo 2 is that Dojo 3 is going to ship a lot faster than Dojo 2 and we'll find ourselves needing to do a major doc update. So maybe express things without a specific version number?